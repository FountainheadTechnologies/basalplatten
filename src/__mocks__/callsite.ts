module.exports = () => ([{
}, {
  getFileName: () => 'test-caller.js'
}]);
// @hack: Make module look like es6 to prevent mangling
module.exports.__esModule = true;
