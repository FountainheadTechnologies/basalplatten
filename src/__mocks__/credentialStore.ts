export const credentialStore = {
  setToken: jest.fn(),
  getToken: jest.fn(),
  clearToken: jest.fn()
}
