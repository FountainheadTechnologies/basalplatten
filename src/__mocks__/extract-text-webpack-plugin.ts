module.exports = class ExtractTextPlugin {
  extract() {}
};
// @hack: Make module look like es6 to prevent mangling
module.exports.__esModule = true;
