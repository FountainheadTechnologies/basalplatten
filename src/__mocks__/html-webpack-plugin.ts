module.exports = class HtmlWebpackPlugin {};
// @hack: Make module look like es6 to prevent mangling
module.exports.__esModule = true;
