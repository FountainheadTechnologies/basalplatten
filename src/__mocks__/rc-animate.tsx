import React from 'react';

const RcAnimate = props => (
  <div>{props.component}</div>
)

RcAnimate.displayName = 'RcAnimate';

export default RcAnimate;
