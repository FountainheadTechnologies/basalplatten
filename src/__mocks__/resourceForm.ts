const submit = jest.fn();

export const createToken = {
  schema: {
    required: ['username', 'password'],
    properties: {
      username: { type: 'string' },
      password: { type: 'string' }
    },
  },

  submit
}

export const createPerson = {
  schema: {
    required: ['forename', 'surname'],
    properties: {
      forename: { type: 'string' },
      surname: { type: 'string' },
      date_of_birth: {
        type: 'string',
        format: 'date-time'
      }
    }
  },

  submit
}

export const updatePerson = {
  schema: {
    required: [],
    default: {
      forename: 'Test',
      surname: 'Person',
      date_of_birth: '1985-08-02T23:00:00.000Z'
    },
    properties: {
      forename: { type: 'string' },
      surname: { type: 'string' },
      date_of_birth: {
        type: 'string',
        format: 'date-time'
      }
    }
  }
}
