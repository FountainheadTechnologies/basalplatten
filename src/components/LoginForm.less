@import "~antd/lib/style/themes/default";

.login-form {
    width: 320px;

    &__alert {
        margin-bottom: @form-item-margin-bottom;
    }

    &__submit-button {
        width: 100%;
    }
}