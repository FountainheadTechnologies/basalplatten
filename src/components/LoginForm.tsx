import React from 'react';
import PropTypes from 'prop-types';
import { Resource, Form as HALForm } from '@optics/hal-client';

import Alert, { AlertProps } from 'antd/lib/alert';
import 'antd/lib/alert/style';
import Button from 'antd/lib/button';
import { ButtonProps } from 'antd/lib/button/button';
import 'antd/lib/button/style';

import { Form as FormComponent } from './Form';
import { SchemaFields } from './SchemaFields';

import './LoginForm.less';

export interface Props {
  resourceForm: HALForm;
  onSuccess?: (resource: Resource) => void;
  alert?: (err?: Error) => Alert | undefined;
  submitButton?: ButtonProps | false;
  fields?: React.StatelessComponent<any> | React.ComponentClass<any> | React.ClassicComponentClass<any>;
}

export interface State {
  error?: Error
}

export class LoginForm extends React.Component<Props> {
  static contextTypes: React.ValidationMap<any> = {
    router: PropTypes.any
  }

  state: State = {};

  render() {
    const {
      resourceForm,
      fields: Fields = SchemaFields
    } = this.props;

    return (
      <div className="login-form">
        {this._renderAlert()}
        <FormComponent
          resourceForm={resourceForm}
          submitButton={this._renderSubmitButton()}
          onSuccess={this._onSuccess}
          onFailure={this._onFailure}
        >
          <Fields />
        </FormComponent>
      </div>
    );
  }

  protected _renderAlert() {
    if (this.props.alert) {
      return this.props.alert(this.state.error);
    }

    const props: AlertProps = this.state.error ? {
      className: 'login-form__alert login-form__alert--error',
      message: this.state.error.message,
      type: 'error',
      showIcon: true
    } : {
        className: 'login-form__alert login-form__alert--info',
        message: 'Please log in to continue',
        type: 'info',
        showIcon: true
      };

    return (
      <Alert {...props} />
    );
  }

  protected _renderSubmitButton() {
    if (this.props.submitButton === false) {
      return;
    }

    const props: ButtonProps = {
      ...this.props.submitButton,
      className: 'login-form__submit-button',
      type: 'primary',
      htmlType: 'submit'
    }

    return (
      <Button {...props}>
        Log In
      </Button>
    );
  }

  protected _onSuccess = (resource: Resource) => {
    if (this.props.onSuccess) {
      this.props.onSuccess(resource);
    }

    this.setState({ error: undefined });
  }

  protected _onFailure = (error: Error) => {
    this.setState({ error });
  }
}
