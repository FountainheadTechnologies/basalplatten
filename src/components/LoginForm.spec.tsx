import React from 'react';
import { mount, ReactWrapper } from 'enzyme';
import { AlertProps } from 'antd/lib/alert';

import { LoginForm } from './LoginForm';
import { createToken } from '../__mocks__/resourceForm';

describe('default behaviour', () => {
  let loginForm: ReactWrapper;
  const onSuccess = jest.fn();

  beforeEach(() => {
    Date.now = jest.fn(() => 100);
    loginForm = mount(
      <LoginForm
        resourceForm={createToken}
        onSuccess={onSuccess}
      />
    );
  });

  it('renders an empty Form with an alert and Log In button', () => {
    expect(loginForm).toMatchSnapshot();
  });

  describe('when `state.error` is set', () => {
    beforeEach(() => {
      loginForm.setState({
        error: Error('oh dear')
      })
    });

    it('displays an Error alert', () => {
      const alertProps = loginForm.find('Alert').props() as AlertProps;

      expect(alertProps.type).toBe('error');
      expect(alertProps.message).toBe('oh dear');
    });
  });

  describe('when form submission succeeds', () => {
    it('calls `onSuccess` prop', () => {
      const result = { token: 'success' };

      loginForm.find('FormComponent').props().onSuccess(result);
      expect(onSuccess).toHaveBeenCalledWith(result);
    });

    it('resets `state.error`', () => {
      loginForm.setState({
        error: Error('oh dear')
      });

      loginForm.find('FormComponent').props().onSuccess();
      expect(loginForm.state('error')).toBeUndefined();
    });
  });

  describe('when form submission fails', () => {
    it('sets `state.error` to the error received', () => {
      const error = Error('oh no');

      loginForm.find('FormComponent').props().onFailure(error);
      expect(loginForm.state('error')).toBe(error);
    });
  });
});

describe('when `props.alert` is a function', () => {
  let loginForm: ReactWrapper;
  let renderAlert = jest.fn();

  beforeEach(() => {
    loginForm = mount(
      <LoginForm
        resourceForm={createToken}
        alert={renderAlert}
      />
    );
  });

  it('is called with `state.error`', () => {
    const error = Error('whoops');

    loginForm.setState({ error });
    expect(renderAlert).toHaveBeenCalledWith(error);
  });

  it('renders the element that was returned', () => {
    renderAlert.mockImplementationOnce(() => (
      <h1>TOP SECRET AREA</h1>
    ));

    loginForm.update();

    expect(loginForm.find('h1').text()).toBe('TOP SECRET AREA');
  });
});

describe('when `props.submitButton` is false', () => {
  let loginForm: ReactWrapper;

  beforeEach(() => {
    loginForm = mount(
      <LoginForm
        resourceForm={createToken}
        submitButton={false}
      />
    );
  });

  it('omits the Submit button', () => {
    expect(loginForm.find('Button').exists()).toBe(false);
  });
});
