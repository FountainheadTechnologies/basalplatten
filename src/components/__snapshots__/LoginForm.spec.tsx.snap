// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`default behaviour renders an empty Form with an alert and Log In button 1`] = `
<LoginForm
  onSuccess={[Function]}
  resourceForm={
    Object {
      "schema": Object {
        "properties": Object {
          "password": Object {
            "type": "string",
          },
          "username": Object {
            "type": "string",
          },
        },
        "required": Array [
          "username",
          "password",
        ],
      },
      "submit": [Function],
    }
  }
>
  <div
    className="login-form"
  >
    <Alert
      className="login-form__alert login-form__alert--info"
      message="Please log in to continue"
      showIcon={true}
      type="info"
    >
      <RcAnimate
        component=""
        onEnd={[Function]}
        showProp="data-show"
        transitionName="ant-alert-slide-up"
      >
        <div />
      </RcAnimate>
    </Alert>
    <Form(WrappedComponent)
      onFailure={[Function]}
      onSuccess={[Function]}
      resourceForm={
        Object {
          "schema": Object {
            "properties": Object {
              "password": Object {
                "type": "string",
              },
              "username": Object {
                "type": "string",
              },
            },
            "required": Array [
              "username",
              "password",
            ],
          },
          "submit": [Function],
        }
      }
      submitButton={
        <Button
          className="login-form__submit-button"
          clicked={false}
          ghost={false}
          htmlType="submit"
          loading={false}
          prefixCls="ant-btn"
          type="primary"
        >
          Log In
        </Button>
      }
    >
      <Component
        form={
          Object {
            "getFieldDecorator": [Function],
            "getFieldError": [Function],
            "getFieldInstance": [Function],
            "getFieldProps": [Function],
            "getFieldValue": [Function],
            "getFieldsError": [Function],
            "getFieldsValue": [Function],
            "isFieldTouched": [Function],
            "isFieldValidating": [Function],
            "isFieldsTouched": [Function],
            "isFieldsValidating": [Function],
            "isSubmitting": [Function],
            "resetFields": [Function],
            "setFields": [Function],
            "setFieldsInitialValue": [Function],
            "setFieldsValue": [Function],
            "submit": [Function],
            "validateFields": [Function],
            "validateFieldsAndScroll": [Function],
          }
        }
        onFailure={[Function]}
        onSuccess={[Function]}
        resourceForm={
          Object {
            "schema": Object {
              "properties": Object {
                "password": Object {
                  "type": "string",
                },
                "username": Object {
                  "type": "string",
                },
              },
              "required": Array [
                "username",
                "password",
              ],
            },
            "submit": [Function],
          }
        }
        submitButton={
          <Button
            className="login-form__submit-button"
            clicked={false}
            ghost={false}
            htmlType="submit"
            loading={false}
            prefixCls="ant-btn"
            type="primary"
          >
            Log In
          </Button>
        }
        wrappedComponentRef={[Function]}
      >
        <FormComponent
          form={
            Object {
              "getFieldDecorator": [Function],
              "getFieldError": [Function],
              "getFieldInstance": [Function],
              "getFieldProps": [Function],
              "getFieldValue": [Function],
              "getFieldsError": [Function],
              "getFieldsValue": [Function],
              "isFieldTouched": [Function],
              "isFieldValidating": [Function],
              "isFieldsTouched": [Function],
              "isFieldsValidating": [Function],
              "isSubmitting": [Function],
              "resetFields": [Function],
              "setFields": [Function],
              "setFieldsInitialValue": [Function],
              "setFieldsValue": [Function],
              "submit": [Function],
              "validateFields": [Function],
              "validateFieldsAndScroll": [Function],
            }
          }
          onFailure={[Function]}
          onSuccess={[Function]}
          resourceForm={
            Object {
              "schema": Object {
                "properties": Object {
                  "password": Object {
                    "type": "string",
                  },
                  "username": Object {
                    "type": "string",
                  },
                },
                "required": Array [
                  "username",
                  "password",
                ],
              },
              "submit": [Function],
            }
          }
          submitButton={
            <Button
              className="login-form__submit-button"
              clicked={false}
              ghost={false}
              htmlType="submit"
              loading={false}
              prefixCls="ant-btn"
              type="primary"
            >
              Log In
            </Button>
          }
          wrappedComponentRef={[Function]}
        >
          <Form
            hideRequiredMark={false}
            layout="horizontal"
            onSubmit={[Function]}
            prefixCls="ant-form"
          >
            <form
              className="ant-form ant-form-horizontal"
              onSubmit={[Function]}
            >
              <Component>
                <div
                  className="schema-fields"
                >
                  <Component
                    key="username"
                    label="Username"
                    name="username"
                  >
                    <FormItem
                      colon={true}
                      hasFeedback={false}
                      key="username"
                      label="Username"
                      prefixCls="ant-form"
                    >
                      <Row
                        className="ant-form-item"
                        gutter={0}
                      >
                        <div
                          className="ant-row ant-form-item"
                        >
                          <Col
                            className="ant-form-item-label"
                            key=".$label"
                          >
                            <div
                              className="ant-form-item-label"
                            >
                              <label
                                className="ant-form-item-required"
                                htmlFor="username"
                                onClick={[Function]}
                                title="Username"
                              >
                                Username
                              </label>
                            </div>
                          </Col>
                          <Col
                            className="ant-form-item-control-wrapper"
                            key=".$wrapper"
                          >
                            <div
                              className="ant-form-item-control-wrapper"
                            >
                              <div
                                className="ant-form-item-control "
                              >
                                <Input
                                  data-__meta={
                                    Object {
                                      "initialValue": undefined,
                                      "leadingName": "username",
                                      "name": "username",
                                      "originalProps": Object {
                                        "disabled": false,
                                        "prefixCls": "ant-input",
                                        "type": "text",
                                      },
                                      "ref": null,
                                      "rules": Array [
                                        Object {
                                          "message": "Username is required",
                                          "required": true,
                                        },
                                      ],
                                      "trigger": "onChange",
                                      "validate": Array [
                                        Object {
                                          "rules": Array [
                                            Object {
                                              "message": "Username is required",
                                              "required": true,
                                            },
                                          ],
                                          "trigger": Array [
                                            "onChange",
                                          ],
                                        },
                                      ],
                                      "valuePropName": "value",
                                    }
                                  }
                                  disabled={false}
                                  id="username"
                                  key=".0"
                                  onChange={[Function]}
                                  prefixCls="ant-input"
                                  size="large"
                                  type="text"
                                >
                                  <input
                                    className="ant-input ant-input-lg"
                                    data-__meta={
                                      Object {
                                        "initialValue": undefined,
                                        "leadingName": "username",
                                        "name": "username",
                                        "originalProps": Object {
                                          "disabled": false,
                                          "prefixCls": "ant-input",
                                          "type": "text",
                                        },
                                        "ref": null,
                                        "rules": Array [
                                          Object {
                                            "message": "Username is required",
                                            "required": true,
                                          },
                                        ],
                                        "trigger": "onChange",
                                        "validate": Array [
                                          Object {
                                            "rules": Array [
                                              Object {
                                                "message": "Username is required",
                                                "required": true,
                                              },
                                            ],
                                            "trigger": Array [
                                              "onChange",
                                            ],
                                          },
                                        ],
                                        "valuePropName": "value",
                                      }
                                    }
                                    disabled={false}
                                    id="username"
                                    onChange={[Function]}
                                    onKeyDown={[Function]}
                                    size="large"
                                    type="text"
                                    value=""
                                  />
                                </Input>
                              </div>
                            </div>
                          </Col>
                        </div>
                      </Row>
                    </FormItem>
                  </Component>
                  <Component
                    key="password"
                    label="Password"
                    name="password"
                  >
                    <FormItem
                      colon={true}
                      hasFeedback={false}
                      key="password"
                      label="Password"
                      prefixCls="ant-form"
                    >
                      <Row
                        className="ant-form-item"
                        gutter={0}
                      >
                        <div
                          className="ant-row ant-form-item"
                        >
                          <Col
                            className="ant-form-item-label"
                            key=".$label"
                          >
                            <div
                              className="ant-form-item-label"
                            >
                              <label
                                className="ant-form-item-required"
                                htmlFor="password"
                                onClick={[Function]}
                                title="Password"
                              >
                                Password
                              </label>
                            </div>
                          </Col>
                          <Col
                            className="ant-form-item-control-wrapper"
                            key=".$wrapper"
                          >
                            <div
                              className="ant-form-item-control-wrapper"
                            >
                              <div
                                className="ant-form-item-control "
                              >
                                <Input
                                  data-__meta={
                                    Object {
                                      "initialValue": undefined,
                                      "leadingName": "password",
                                      "name": "password",
                                      "originalProps": Object {
                                        "disabled": false,
                                        "prefixCls": "ant-input",
                                        "type": "text",
                                      },
                                      "ref": null,
                                      "rules": Array [
                                        Object {
                                          "message": "Password is required",
                                          "required": true,
                                        },
                                      ],
                                      "trigger": "onChange",
                                      "validate": Array [
                                        Object {
                                          "rules": Array [
                                            Object {
                                              "message": "Password is required",
                                              "required": true,
                                            },
                                          ],
                                          "trigger": Array [
                                            "onChange",
                                          ],
                                        },
                                      ],
                                      "valuePropName": "value",
                                    }
                                  }
                                  disabled={false}
                                  id="password"
                                  key=".0"
                                  onChange={[Function]}
                                  prefixCls="ant-input"
                                  size="large"
                                  type="text"
                                >
                                  <input
                                    className="ant-input ant-input-lg"
                                    data-__meta={
                                      Object {
                                        "initialValue": undefined,
                                        "leadingName": "password",
                                        "name": "password",
                                        "originalProps": Object {
                                          "disabled": false,
                                          "prefixCls": "ant-input",
                                          "type": "text",
                                        },
                                        "ref": null,
                                        "rules": Array [
                                          Object {
                                            "message": "Password is required",
                                            "required": true,
                                          },
                                        ],
                                        "trigger": "onChange",
                                        "validate": Array [
                                          Object {
                                            "rules": Array [
                                              Object {
                                                "message": "Password is required",
                                                "required": true,
                                              },
                                            ],
                                            "trigger": Array [
                                              "onChange",
                                            ],
                                          },
                                        ],
                                        "valuePropName": "value",
                                      }
                                    }
                                    disabled={false}
                                    id="password"
                                    onChange={[Function]}
                                    onKeyDown={[Function]}
                                    size="large"
                                    type="text"
                                    value=""
                                  />
                                </Input>
                              </div>
                            </div>
                          </Col>
                        </div>
                      </Row>
                    </FormItem>
                  </Component>
                </div>
              </Component>
              <Button
                className="login-form__submit-button"
                clicked={false}
                ghost={false}
                htmlType="submit"
                loading={false}
                prefixCls="ant-btn"
                type="primary"
              >
                <button
                  className="ant-btn login-form__submit-button ant-btn-primary"
                  onClick={[Function]}
                  type="submit"
                >
                  <span
                    key=".0"
                  >
                    Log In
                  </span>
                </button>
              </Button>
            </form>
          </Form>
        </FormComponent>
      </Component>
    </Form(WrappedComponent)>
  </div>
</LoginForm>
`;
